### PR TITLE
Add Bow histogram computation

### PR DESCRIPTION
**Pull request by ovysotska**
This PR adds the Python code to:
* train the bag of words vocabulary
* extract image histograms based on pre-trained bag of words vocabulary
* some tests :smile: 

In the testing mode, the code takes a folder with images and returns the np.array of Bow features stored in the specified file.

The computed features can be used in the current codebase further, by first turning them into proto files, and then the full pipeline for image_sequence_localizer can be used as usual.

### CODE REVIEW

**Comment by ovysotska** [comment]
Hey @niosus  would you have time to review this one?

File: src/python/bow/bow.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+import cv2
+import argparse
+from pathlib import Path
+import matplotlib.pyplot as plt
+
+from sklearn import preprocessing
+from sklearn.cluster import KMeans
+from sklearn.neighbors import KDTree
+
+kDefaultWidth = 640  # px
```
Comment (niosus) [inline src/python/bow/bow.py:11]: Love the names for the constants 😍

File: src/python/bow/bow.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+kDefaultClusterSize = 400
+
+
+def listImagesInFolder(folderPath):
+    trainImageFiles = list(folderPath.glob("*.jpg"))
+    trainImageFiles.extend(list(folderPath.glob("*.png")))
+    return trainImageFiles
+
+
+def rescaleImageIfNeeded(image):
```
Comment (niosus) [inline src/python/bow/bow.py:21]: nit: very minor, but you seem to have docs for the other functions and not for these. Maybe add docstring to these too?

File: src/python/bow/bow.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    # Extracts features from an image
+    image = cv2.imread(imageFile.as_posix(), cv2.IMREAD_GRAYSCALE)
+    image = rescaleImageIfNeeded(image)
+    sift = cv2.SIFT_create()
+    keypoints, descriptors = sift.detectAndCompute(image, None)
+    return descriptors
+
+
+def computeIDF(descriptorsPerImage, clusters):
+    """Compute inverse document frequence (IDF). Here means in how many images does the word occur.
```
Comment (niosus) [inline src/python/bow/bow.py:48]: > Here means in how many images does the word occur.

nit: This reads a bit strange. What did you mean to say here?

File: src/python/bow/bow.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    descriptorsPerImage = []
+    for imageFile in imageFiles:
+        sifts = extractSiftsFromImage(imageFile)
+        descriptorsPerImage.append(sifts)
+
+    # flatten the descriptors list
+    descriptors = [
+        descriptor for descriptors in descriptorsPerImage for descriptor in descriptors
+    ]
+    descriptors = np.array(descriptors)
```
Comment (niosus) [inline src/python/bow/bow.py:92]: I don't really know how to improve this, but the naming seems to be overloaded here. The word `descriptors` is very overused here. Maybe use more descriptive names?

File: src/python/bow/bow.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+    descriptorsNormalized = preprocessing.normalize(descriptors)
+    kmeans = KMeans(n_clusters=kDefaultClusterSize, random_state=0, n_init="auto")
+    kmeans.fit(descriptorsNormalized)
+    words = kmeans.cluster_centers_
+
+    idfs = computeIDF(descriptorsPerImage, words)
+
+    plt.bar(range(0, len(idfs)), idfs)
+    plt.savefig("idf_" + str(kDefaultClusterSize) + ".png")
```
Comment (niosus) [inline src/python/bow/bow.py:102]: nit: Do you want to always save these images? Or should it be hidden behind some flag?

File: src/python/bow/bow.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        print("Vocabulary was saved to", outputFile)
+    return words, idfs
+
+
+def trainVocabularyFromFolder(folderPath, outputFile=""):
+    return trainVocabulary(listImagesInFolder(folderPath), outputFile)
+
+
+def getVocabulary(imageTrainFolder, vocabularyFile):
+    if vocabularyFile is not None:
```
Comment (niosus) [inline src/python/bow/bow.py:115]: nit: you can simplify this to:
```python
if vocabularyFile and vocabularyFile.exists():
```

**Review by niosus — APPROVED** [review]
A solid PR, @ovysotska! I left a couple of nitpicks here and there, but feel free to ignore them. I think the code is easy to read and the intent is clear. Added plus for having a couple of tests.

File: src/python/bow/bow.py (hunk 3 above)
Comment (ovysotska) [reply src/python/bow/bow.py:48]: I meant the IDF (which is inverse document frequency) , here stands for how many images (not documents) does the word occur in.

File: src/python/bow/bow.py (hunk 4 above)
Comment (ovysotska) [reply src/python/bow/bow.py:92]: It is the best. I also don't know now what that suppose to be :)

File: src/python/bow/bow.py (hunk 5 above)
Comment (ovysotska) [reply src/python/bow/bow.py:102]: Neh, not really important. Will just remove it and reduce the dependencies list.

**Pull request merged**